### ISSUE TITLE
false diagnostics with -fopenmp

### ISSUE THREAD

**Issue by scturtle** [MaskRay/ccls#113]
Latest ccls compiled with LLVM 7. `-fopenmp` causes false diagnostics.
```c++
#include <Eigen/Core>
#include <glog/logging.h>
int main() {LOG(INFO);} //  ccls: use of undeclared identifier 'INFO' (lsp-ui)
```
I just remove `-fopenmp` from `args` manually at the time. May not be a good solution.

**Comment by MaskRay** [comment]
@acowley seems to have the same issue. I'll investigate what `-fopenmp` does

**Comment by acowley** [comment]
Yes, I add it as a filtered flag to the ccls startup options. The only downside is warnings about unknown openmp pragmas in my code.

**Comment by MaskRay** [comment]
I cannot reproduce the spurious diagnostic with 

```zsh
hg clone https://bitbucket.org/eigen/eigen

git clone https://github.com/google/glog
cd glog; ./autogen.sh; ./configure; make -j
```

and `-Ieigen -Iglog/src -fopenmp` ...

**Comment by scturtle** [comment]
`cd eigen; hg up 3.3.5`. The latest stable release version of eigen should be able to reproduce it. There should be a bug in it that frustrates ccls a lot.

**Comment by MaskRay** [comment]
Sorry I can reproduce it, with either head or 3.3.5

It is an issue of preamble building with `-fopenmp`.

In `clang_complete.cc:DiagnosticMain`, if I don't use preamble (basically longest prefix of preprocessor tokens), the spurious error goes away.

```diff
-    std::shared_ptr<PreambleData> preamble = session->GetPreamble();
+    std::shared_ptr<PreambleData> preamble;
```

When building the preamble.i n `clang/lib/Parse/ParseOpenMP.cpp ParseOpenMPDeclarativeDirectiveWithExtDecl`, this is reported as a `diag::err_omp_unexpected_directive` because `Tok` is `clang:;tok::TokenKind::identifier`.

**Comment by MaskRay** [comment]
@Riatre I feel this same. When I get more time I'll study more about clang's lexer/parser.

**Comment by amosbird** [comment]
another MVE for this

```
#include <vector>

#include <parallel/algorithm>

int main() {
    std::vector<int> a = { 3, 2, 1 };
    __gnu_parallel::sort(std::begin(a), std::end(a));
    return 0;
}
```

**Comment by bastianbeischer** [comment]
@MaskRay @Riatre This is still an open issue with LLVM 8 and latest ccls for me. As far as I understand this is a LLVM problem and not directly related to ccls. So is there a bug report for LLVM/clang to track this?

**Comment by vvhitedog** [comment]
FYI this is still very much an issue in LLVM 9.0.1, with the latest ccls at time of writing. Turning off openmp in a project that utilizes Eigen fixes all the false diagnostics.
